### PR TITLE
New concept exercise `dancing-dots` (`use` and `behaviour`)

### PR DESCRIPTION
**Pull request by angelikatyborska**
~Heavily WIP!~

~I'm opening this PR to signal that I'm working on this.~

This will be necessary as a prerequisite for the new GenServer exercise (https://github.com/exercism/elixir/pull/1076)

### CODE REVIEW

File: exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex
Comment (angelikatyborska) [inline exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex:4]: I'm not sure if it's a good idea to leave this code here. But one of my biggest problems with understanding behaviours is actually imagining how they can be used once I have some modules that all implement that one behaviour.

**Comment by angelikatyborska** [comment]
@jiegillet The example solution, tests, and instructions are ready for review. Can you take a look?

File: exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex
Comment (jiegillet) [reply exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex:4]: Yeah, I'm with you, it's nice to see how it can be used. `render_dots/2` is great because it's very readable, but I think maybe `new/2` is a bit too complex... Maybe it could be changed to filter the valid options rather than throwing an error for the sake of simplicity?

(untested idea)
```elixir
def new(dots, animations_with_opts) do
  valid_animations_with_opts =
  for {animation_module, opts} <- animations_with_opts,
      # using Animation's init/1 callback
      {:ok, opts} <- animation_module.init(opts) do
    {animation_module, opts}
  end

  {:ok,  %__MODULE__{dots: dots, animations_with_opts: valid_animations_with_opts }}
end

```

File: exercises/concept/dancing-dots/.meta/design.md
Comment (jiegillet) [inline exercises/concept/dancing-dots/.meta/design.md:10]: -  `@macrocallback`
- `@optional_callbacks`

File: exercises/concept/dancing-dots/.meta/exemplar.ex
Comment (jiegillet) [inline exercises/concept/dancing-dots/.meta/exemplar.ex:39]: I would suggest to add one type of errors here: "Missing option :velocity". And some tests for that.

File: exercises/concept/dancing-dots/test/dancing_dots/animation_test.exs
Comment (jiegillet) [inline exercises/concept/dancing-dots/test/dancing_dots/animation_test.exs:173]: How about splitting this test in two here and mention negative velocity? Just because the instructions mention it but the test descriptions don't.

**Review by jiegillet — COMMENTED** [review]
Very very cool! The story is great and there is a lot to learn.

I took my time solving with only the instructions and the Elixir documentation. What a mess!! (the docs, not your instructions) All the ingredients you need for behaviours are scattered everywhere in like 3 or 4 different pages. This really is a needed exercise! I think even a nice blog post about the topic would be great.

For reference, here is my final submission, it's very very similar to yours:
```elixir
defmodule DancingDots.Animation do
  @type dot :: DancingDots.Dot.t()
  @type opts :: keyword
  @type error :: any 
  @type frame_number :: pos_integer

  @callback init(opts) :: {:ok, opts} | {:error, error}
  defmacro __using__(_opts) do
    quote do
      @behaviour DancingDots.Animation
      def init(opts) do
        {:ok, opts}
      end 

      defoverridable init: 1
    end 
  end 

  @callback handle_frame(dot, frame_number, opts) :: dot 
end

defmodule DancingDots.Flicker do
  use DancingDots.Animation
  alias DancingDots.Dot

  @impl DancingDots.Animation
  def handle_frame(%Dot{opacity: opacity} = dot, frame_number, _opts)
      when rem(frame_number, 4) == 0 do
    %Dot{dot | opacity: opacity / 2}
  end 

  def handle_frame(%Dot{} = dot, _frame_number, _opts), do: dot 
end

defmodule DancingDots.Zoom do
  use DancingDots.Animation
  alias DancingDots.Dot

  @impl DancingDots.Animation
  def init(opts) do
    velocity = Keyword.get(opts, :velocity)

    if is_number(velocity) do
      {:ok, opts}
    else
      {:error, "Expected required option :velocity to be a number, got: #{inspect(velocity)}"}
    end 
  end 

  @impl DancingDots.Animation
  def handle_frame(%Dot{radius: radius} = dot, frame_number, opts) do
    n = frame_number - 1 
    velocity = Keyword.get(opts, :velocity)
    %Dot{dot | radius: radius + n * velocity}
  end 
end
```

Outside of finding information in the docs, the only place that gave me pause was the error message in `DancingDots.Zoom`, I left some comments about that.

File: exercises/concept/dancing-dots/.meta/exemplar.ex
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.meta/exemplar.ex:39]: I'm not sure I understand the suggestion, are you saying there need to be two different errors, one for when velocity was not given, and one for when it's given but not valid? There already is a test for when it's missing:
```elixir
      assert DancingDots.Zoom.init([]) ==
               {:error, "Expected required option :velocity to be a number, got: nil"}
```

File: exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex:4]: I agree, `new` was too complex. I think the worst offender there was the `reduce_while` call. To make it simpler, I decided to split `new` into two functions - `new` that just returns a group with no animations, and `add_animation` to add a single animation module to the list. That allows getting rid of any `Enum` calls.

I didn't want to silently ignore initialization errors in this demo because that would show an anti-pattern IMO.

File: concepts/behaviours/about.md
Comment (angelikatyborska) [inline concepts/behaviours/about.md:3]: TODO before merging

File: concepts/behaviours/introduction.md
Comment (angelikatyborska) [inline concepts/behaviours/introduction.md:3]: TODO before merging

File: concepts/use/about.md
Comment (angelikatyborska) [inline concepts/use/about.md:3]: TODO before merging

**Comment by angelikatyborska** [comment]
This is now ready for the final full review.

File: exercises/concept/dancing-dots/.docs/introduction.md
Comment (kotp) [inline exercises/concept/dancing-dots/.docs/introduction.md:39]: Same here for English (US) spelling.

File: exercises/concept/dancing-dots/.docs/introduction.md
Comment (kotp) [inline exercises/concept/dancing-dots/.docs/introduction.md:45]: Even if we accept that this exercise is the exception to the US English spelling of words (or perhaps "behaviours" is the only exception in this exercise) then it should at least be consistently spelled.

Comment (kotp) [reply exercises/concept/dancing-dots/.docs/introduction.md:45]: Sorry, this is all I have time for now…

File: exercises/concept/dancing-dots/.docs/hints.md
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.docs/hints.md:7]: Thanks, but in Elixir, "behaviour" is a concept and whenever we refer to this concept, we need to use the UK spelling because it affects the code too (e.g. you need to write `@behaviour DancingDots.Animation`, not `@behavior`). Other times, when we use "behavior" in its regular life non-Elixir meaning, we use the US spelling.

There was one mistake in my PR, I fixed it: [`5827335` (#1103)](https://github.com/exercism/elixir/pull/1103/commits/5827335fb684a5c9fe7914f6f09efcc84d7a1909)

File: exercises/concept/dancing-dots/.docs/hints.md
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.docs/hints.md:33]: We use the real function/macro names in link names, hence `rem`.

File: exercises/concept/dancing-dots/.docs/hints.md
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.docs/hints.md:30]: I disagree, this makes it much harder for us to know where the link leads to. The text "this one special macro" is cryptic on purpose, to hide from the students what it is, but as a maintainer I want to see immediately that this link leads to the documentation about "use".

File: exercises/concept/dancing-dots/.docs/instructions.md
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.docs/instructions.md:5]: The word was duplicated on purpose as a way to straighten its meaning. Does that need a comma? I'm not the best with punctuation... Anyway, I'll just remove the duplication.
```suggestion
Constraints help creativity and shorten project deadlines, so you've both agreed to limit your masterpiece to a single shape - the circle. But there's going to be many circles. And they can move around! You'll call it... dancing dots.
```

File: exercises/concept/dancing-dots/.docs/instructions.md
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.docs/instructions.md:60]: Thanks! Your suggestion does sound much better.

File: exercises/concept/dancing-dots/.meta/exemplar.ex
Comment (jiegillet) [reply exercises/concept/dancing-dots/.meta/exemplar.ex:39]: Yes, I was suggesting two different errors, but actually a simple rephrasing might do.

If no `:velocity` option is given, the error message says "Expected required option `:velocity` to be a number", which feels like you passed something wrong, not nothing at all.

How about:
"The `:velocity` option is required, and its value must be a number. Got: nil"
It's a small change, but it feels a bit better maybe?

File: concepts/behaviours/.meta/config.json
Comment (jiegillet) [inline concepts/behaviours/.meta/config.json:2]:
```suggestion
  "blurb": "Behaviours allow us to define interfaces in a behaviour module that can be later implemented by different callback modules.",
```

File: exercises/concept/dancing-dots/.docs/introduction.md
Comment (jiegillet) [inline exercises/concept/dancing-dots/.docs/introduction.md:84]: This is a bit tricky to figure out, so maybe being a bit more precise would help?

```suggestion
When defining a behaviour, it is possible to provide a default implementation of a callbacks. This implementation should be defined in the quoted expression of the `__using__/1` macro. To make it possible for users of the behaviour module to override the default implementation, call the `defoverridable/1` macro after the function implementation. It accepts a keyword list of function names as keys and function arities as values.
```

Would it be worth adding an example as well?

```elixir
defmodule Countable do
  @callback count(collection :: any) :: pos_integer
  @callback indexed_from() :: pos_integer

  defmacro __using__(_) do
    quote do
      @behaviour Countable
      def indexed_from(), do: 1
      defoverridable indexed_from: 0
    end 
  end 
end
```

File: exercises/concept/dancing-dots/.meta/design.md
Comment (jiegillet) [inline exercises/concept/dancing-dots/.meta/design.md:26]: Is there a need for an analyzer as well?

The only thing I can thing of is that `DancingDots.Zoom.init/1` uses `Keyword` and `is_number` but it's very minor.

File: exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex
Comment (jiegillet) [reply exercises/concept/dancing-dots/lib/dancing_dots/dot_group.ex:4]: Yes, this new version is great!

**Review by jiegillet — APPROVED** [review]
Awesome work 🎉 
I left some comments, but this is already good enough to merge.

File: exercises/concept/dancing-dots/.meta/exemplar.ex
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.meta/exemplar.ex]: Totally, that does sound better. Here: [`ccfcabe` (#1103)](https://github.com/exercism/elixir/pull/1103/commits/ccfcabeb06e86eda16a84d68ffea071b3b5951f4)

File: exercises/concept/dancing-dots/.meta/design.md
Comment (angelikatyborska) [reply exercises/concept/dancing-dots/.meta/design.md:26]: I'm not sure if those two things even matter, using the access behaviour (`opts[:velocity]`) is also fine. I guess the only "bad idea" that technically works would be pattern matching on the keyword list. As for `is_number`, we could add more tests that would assert that it also allows floats or something, but I don't think that's really necessary.

I could think of two things, relevant to the concept:

[`362ceec` (#1103)](https://github.com/exercism/elixir/pull/1103/commits/362ceeca1b02e8256a11d383a404dd26d6bd9d6f)

Comment (jiegillet) [reply exercises/concept/dancing-dots/.meta/design.md:26]: Those are better, yes :)

File: exercises/concept/dancing-dots/.docs/instructions.md
Comment (kotp) [reply exercises/concept/dancing-dots/.docs/instructions.md]: This is already in, but "there are going to be very many circles" though yes, "many, many" would be the way to strengthen the idea of "a whole bunch of" them.

**Pull request merged**
